### ISSUE TITLE
Return this type for fluent API

### ISSUE THREAD

**Issue by VanKrock** [dotnet/csharplang#902]
Now we can't create true fluent api with inharitance and interfaces

We can use generics:
```csharp
    public interface IMyClass<out TClass> where TClass : IMyClass<TClass>
    {
        TClass SomeChange();
    }

    public class MyClass<TClass> : IMyClass<TClass> where TClass : IMyClass<TClass>
    {
        public TClass SomeChange()
        {
            return (TClass)(IMyClass<TClass>)this;
        }
    }
```
But we have a problems with inheritance of interface and Implicit casts.

We can use extensions:
```csharp
    public interface IMyClass
    {
    }

    public class MyClass : IMyClass
    {
    }

    public static class MyClassExtensions
    {
        public static TClass SomeChange<TClass>(this TClass myClass) where TClass : IMyClass
        {
            return myClass;
        }
    }
```
But we can't use protected properties.

I see useful return this:
```csharp
    public interface IMyClass
    {
        this SomeChange();
    }

    public class MyClass : IMyClass
    {
        public this SomeChange()
        {
        }
    }
```
It's can work like void, but return this

**Comment by svick** [comment]
Related: https://github.com/dotnet/csharplang/issues/311, https://github.com/dotnet/csharplang/issues/252, https://github.com/dotnet/csharplang/issues/169.

**Comment by yaakov-h** [comment]
I'm wondering if we can borrow `instancetype` from Objective-C to solve this problem.

It works by saying "this method or property returns a type of the current instance". For example:

```
interface IA {
   instancetype Foo();
}

class B : IA {
    public instancetype Foo() => ...;
}
```

In this instance, `B.Foo()` would return an object of type `B`.

Of course, Objective-C has a much more dynamic runtime than .NET does, so the following would have to be invalid in C#:

```
class C : B {
}
```

In this instance, `C.Foo` has no way to return an instance of `C`, because the most specific implementation of the method is defined on `B`. `instancetype`/`this` might have to require any method/property to be overridden at every (non-abstract) level in the inheritance tree.

**Comment by jnm2** [comment]
Also possibly related is the override return type covariance proposal?

**Comment by VanKrock** [comment]
@yaakov-h I can't understand how we can work with instancetype. How we should create new instancetype and how we should use returned instatcetype.

**Comment by TonyValenti** [comment]
Perhaps you might want to consider making the protected variables "internal protected".

**Comment by GeraudFabien** [comment]
I don't understand. Why don't you just do this.
```
    interface IMain
    {
        IMain GetA();
        IMain GetB();
        IMain GetC();
    }

    class A: IMain
    {
        public virtual IMain GetA()
        {
            Console.WriteLine("A");
            return this;
        }

        public virtual IMain GetB()
        {
            Console.WriteLine("B");
            return new B();
        }

        public virtual IMain GetC()
        {
            Console.WriteLine("C");
            return new C();
        }
    }

    class B : A
    {
        public override IMain GetA()
        {
            Console.WriteLine("A");
            return new A();
        }

        public override IMain GetB()
        {
            return this;
        }

        public override IMain GetC()
        {
            return base.GetC();
        }
    }

    class C : IMain
    {
        public virtual IMain GetA()
        {
            Console.WriteLine("A");
            return new A();
        }

        public virtual IMain GetB()
        {
            Console.WriteLine("B");
            return new B();
        }

        public virtual IMain GetC()
        {
            Console.WriteLine("C");
            return this;
        }
    }

    class Program
    {
        
        static void Main(string[] args)
        {
            Console.WriteLine((new A()).GetA().GetB().GetC().GetType());
            Console.ReadLine();
        }
    }
```
I mean having one interface for your fluent API. If you change class all your class need to implement the same method. So why don't you just make a common interface you return each time.

**Comment by VanKrock** [comment]
@GeraudFabien how you can add GetD() if IMain is third party library?
 we can't to write:
```csharp
interface IDerived : IMain
{
    IDerived GetD();
}
```

**Comment by yaakov-h** [comment]
Well you can, but then `IDerived.GetA()` will return `IMain`, not `IDerived` and you'd have to cast it every time.

Another example is `IClonable.Clone` - if you clone an object, you'd want to return an object of the type of the current object, not a parent class.

**Comment by VanKrock** [comment]
@yaakov-h IClonable is not another example, because in method Clone() we bust to create new instance. We can't create new instance of this. We can't know in advance constructor of deriverd class. We can change this instance and return this.

**Comment by jnm2** [comment]
@VanKrock `IClonable` is a perfect example. The method is abstract so the derived class is responsible for creating another instance of itself _or_ returning the same instance. The important part is that the return type is known statically to be not `object`, not `BaseClass`, but whatever the most derived class is for that instance.

**Comment by VanKrock** [comment]
@jnm2 It's another logic, it's not this, may be "instancetype" or else. Instancetype cannot be created in current class, must be abstract and must be overrided in each derived class. But "this" - "can work like void, but return this";

**Comment by VanKrock** [comment]
```csharp
    public interface IMyClass
    {
        this SomeChange();
    }

    public class MyClass : IMyClass
    {
        public this SomeChange()
        {
        }
    }

    public class Test
    {
        public void Start(IMyClass myClass)
        {
            myClass.SomeChange().SomeChange();
        }
    }
```

Equals
```csharp
    public interface IMyClass
    {
        void SomeChange();
    }

    public class MyClass : IMyClass
    {
        public void SomeChange()
        {
        }
    }

    public class Test
    {
        public void Start(IMyClass myClass)
        {
            myClass.SomeChange();
            myClass.SomeChange();
        }
    }
```
It's syntax sugar

**Comment by Logerfo** [comment]
## Alternatives
### Alternative 1
```c#
public class B
{
   public virtual B M() => this;
}

public class C : B
{
   public override B M() => this;
}

public static class D
{
   public static C M(this C c) => (C)c.M();
}
```
**Problem**: a cast is necessary at every call.

### Alternative 2
```c#
public class B<T> where T : B<T>
{
   public virtual T M() => (T)this;
}

public class Z<T> : B<T> where T : B<T> { }

public class C : Z<C>
{
   public override C M() => this;
}

public static class D
{
   public static C M(this C c) => c.M();
}
```
**Problem**: it is not possible to have non generic intermediary classes inheriting B (like non generic Z).

##
I think this would be a great feature!

## Bonus
This could also work for extensions methods:
```c#
public static this M(this C c) { }
```

**Comment by VanKrock** [comment]
@Logerfo We can't call extension method M because we have non extension method M with identical signature.

But for extensions you can use:
```csharp
public static T M<T>(this T c) where T : C { return (T)c.SomeMethodReturnC(); }
```

**Comment by Logerfo** [comment]
@VanKrock my bonus example was supposed to be random, I mean, any method name will do the job to avoid the boilerplate you just wrote.

**Comment by VanKrock** [comment]
@Logerfo no your example works only for C it's can be
```csharp
public static C M(this C c) { }
```
but generics is not bad for this task. My example usage:
```csharp
var c = new C();
c = c.M();
```
Extensions works cool, but it's have some problems:
1) We can't to use private and protected properties
2) We can to have several extensions with identical signature

**Comment by mikernet** [comment]
I imagine the way this would be implemented is that if you derived a type from another type that had methods that return `this`, then each derived type would just generate a `new` method that hides the base method and does the cast to its own type since that is guaranteed to work.

**Comment by mikernet** [comment]
```c#
public class A
{
    public this AddSomething(Something something);
    // Above gets compiled to: 
    [ThisReturnAttribute]
    public A AddSomething(Something something);
}

public class B : A
{
    // Auto-generated by compiler:
    public new B AddSomething(Something something) => ((A)this).AddSomething();
}

public class C : B
{
    // Auto-generated by compiler:
    public new C AddSomething(Something something) => ((A)this).AddSomething();
}
```

**Comment by SilentSin** [comment]
> I imagine the way this would be implemented is that if you derived a type from another type that has methods that return this, then each derived type would just generate a new method that hides the base method and does the cast to its own type since that is guaranteed to work.

I imagine a simpler solution where it's just syntax sugar on the caller side, much like extension methods. Currently the only way to implement method chaining is by having all your methods `return this` all the time, but isn't the only way it *could* be done.

`public this AddSomething(Something something)` would compile the return type to `void` and all callers would compile from:

```
void SomeMethod(A a)
{
    a.AddSomething(thing)
        .AddSomething(anotherThing)
        .AddSomething(yetAnotherThing);
}
```

To the standard non-fluent way of doing it:

```
void SomeMethod(A a)
{
    a.AddSomething(thing);
    a.AddSomething(anotherThing);
    a.AddSomething(yetAnotherThing);
}
```

This way the `AddSomething` method doesn't actually need to return a value - even implicitly - and the compiler doesn't need to add duplicate methods containing casts to all child types. The caller already has the object, so if this feature only allows repeated use of that object then there's no need to keep getting it back from every method.

**Comment by HaloFour** [comment]
@SilentSin 

That approach wouldn't work terribly well with immutable builders, like `ImmutableArray<T>`, `String` or anything LINQ.

**Comment by SilentSin** [comment]
@HaloFour 

Perhaps you could elaborate because I don't see the issue.

Converting to an immutable array is obviously not going to be returning the enclosing object type so this feature would never be able to work for such a method, regardless of which approach the implementation uses.

String methods never return the original object so it doesn't apply to them either.

And not many LINQ methods appear to directly return the `this` parameter either.  If for whatever reason they need to have an actual return type which can't just be syntax sugar, then they should continue doing so without mangling this feature just to fit them in. But I can't see how @mikernet's implementation could ever work for them either. A class *could* automatically generate methods to hide those from the base class, but doing it like that for LINQ would require the class to automatically generate methods to hide a particular set of extension methods.

**Comment by mikernet** [comment]
It doesn't have to generate anything. In terms of implementation, it can simply be a `[ThisAttribute]` annotation on the return value which the compiles to a cast in the bytecode to the current object type that the method is called on. The JIT would eliminate the cast as it wouldn't be necessary.

**Comment by mikernet** [comment]
@SilentSin Perhaps you are right in the sense that callsite syntactic sugar is the way to go, but in that case the return value should still be there and it should just auto-generate a cast to the type the method is called on at the callsite. The JIT could eliminate the cast for performance as it isn't actually necessary.

**Comment by HaloFour** [comment]
@SilentSin 

My point is that fluent APIs like this don't always return the same instance, so having the method return `void` and having cascades as a caller-side feature doesn't solve those cases.

**Comment by SilentSin** [comment]
@mikernet But under what circumstances could the return value ever actually be used? A normal call would skip it and a reflection call would already have to have the object reference.

@HaloFour I don't see how fluent APIs that return a different instance could actually make use of this feature though. If @mikernet's `AddSomething` method wanted to create and return a `new A` with the element added, then the return type obviously needs to be `A` and the casts in `B` and `C` would fail.

**Comment by HaloFour** [comment]
@SilentSin 

Many of those APIs return the same type, just a different instance of that type.

**Comment by SilentSin** [comment]
@HaloFour Yes, but while it might be nice for those to be able to avoid naming the return type in the method signature, that is the extent of the benefit such methods could gain from this feature. The real benefit is the way it allows methods in the base class to return the child class when called on an instance of the child class. The only guaranteed reference the base class has to the specific child class is the `this` pointer.

Actually, this feature could easily support both. Use the `this` return type without a `return` statement and it will implicitly `return this` and can be used to chain calls on the child type. Or put in a `return` statement and it will only use the `this` return type as shorthand for `BaseClass`.

**Comment by mikernet** [comment]
You guys are confusing things here.

@HaloFour LINQ methods don't return the same type as the instance - they (usually) return `IEnumerable<T>`. A `List<int> x;` having a LINQ method called on it like `x.Where(i => i.IsGood)` does not return a `List<T>`. 

This is for a difference scenario. This is for things like fluent interface builder methods that always return **this** and only **this**. Otherwise you can't guarantee that the return value for a subclass will be the same type as the object the method is called on.

@SilentSin A normal call would skip what? I don't understand what you are saying. Here's a scenario:

```c#
public class Person
{
    public this SetName(string name); // return this;
}

public class Worker : Person
{
    public this SetWage(decimal amount); // return this;
}

Worker bob = new Worker().SetName("Bob").SetWage(50000);
// compiles to:
Worker bob = ((Worker)new Worker().SetName("Bob")).SetWage(50000);

Worker bill = new Worker().SetName("Bill");
//compiles to:
Worker bill = (Worker)new Worker().SetName("Bill");
```

Because the base class fluent method returns an actual worker and not a person this works. If that wasn't the case, `SetWage()` fails in the first example because `SetName()` returns a person and a person doesn't have a wage. The second example fails because `SetName()` returns a person so it needs to be cast back to a worker before being stored in a worker variable.

**Comment by mikernet** [comment]
@SilentSin I think I see what you were saying before though...if the return is always `this` then why even return it? The first example can just compile to:

```c#
Worker bob = new Worker().SetName("Bob").SetWage(50000);
// compiles to:
Worker bob = new Worker();
worker.SetName("Bob");
worker.SetWage(50000);
```
And it will continue to work...no need for casting. I'm torn by which is better, because at least if it returns the instance then the fluent interface is still somewhat usable in previous C# versions and/or other languages. I don't know how much that matters tbh though.

**Comment by HaloFour** [comment]
@mikernet 

> This is for things like fluent interface builder methods that always return **this** and only **this**.

Not all fluent builders return `this`, they return a new instance of the same type.  `ImmutableArray<T>` is a good example.  You could consider `String` an example as well.  I think it would be short-sighted for a feature like this to try to force that the builder itself must be mutable.

**Comment by mikernet** [comment]
@HaloFour  There are already perfectly good facilities for implementing that behavior. How is this necessary to make that work or make it better?

**Comment by mikernet** [comment]
@HaloFour  The only reason you would need this is if a fluent interface is subclassable. String is not subclassable so it doesn't apply. I also see zero reason to make `ImmutableArray<T>` subclassable.

**Comment by CyrusNajmabadi** [comment]
> String is not subclassable so it doesn't apply. I also see zero reason to make ImmutableArray<T> subclassable.

It would apply to anyone who *does* have a subclassable fluent interface.

**Comment by HaloFour** [comment]
@mikernet 

Because `this` type is not exclusive to mutable builders with subclassing.  It's related to contravariant returns which is applicable to any form of fluent API.

**Comment by mikernet** [comment]
What?!? Rofl I have no idea what you guys (@CyrusNajmabadi @HaloFour) are talking about.

Those were examples @HaloFour gave, which are not subclassable so they aren't good examples. Give me an example of something that is subclassable that should return an instance of whatever type it is called on because I can't think of any. Give me a concrete example of what you are trying to achieve that you can't today.

**Comment by CyrusNajmabadi** [comment]
> What?!? Rofl I have no idea what you guys (@CyrusNajmabadi @HaloFour) are talking about.

It's very simple:


```c#
class MyBuilder {
     DerivedBuilderForSpecificPurpose SomeFluentCall();
}

class DerivedBuilderForSpecificPurpose  : MyBulder {
     MyBuilder DerivedBuilderSpecificMember1() { ... }
     DerivedBuilderForSpecificPurpose  DerivedBuilderSpecificMember2();
     AnotherDerivedBuilderForAnotherPurpose DerivedBuilderSpecificMember3();
}
```

Etc. etc.  This is totally normal as you may often perform a fluent step that returns another view of things that provides different functionality based on the step you just performed.  A trivial example of this is `.OrderBy` in the linq space.  Once you call that, you get an IOrderedEnumerable which then exposes additional extensions like '.ThenBy'.

**Comment by mikernet** [comment]
I don't think you are considering how what you are asking is impossible.

Let's say you have IOrderedEnumerable...okay, so now how does LINQ method `Where` called on an `IOrderedEnumerable` return an `IOrderedEnumerable` instance? It can't. Or what is `Where` was called on a `List<T>`? How do you write the `Where()` method so that it returns the same type that it is called on?

**Comment by mikernet** [comment]
How do you write the method in your BASE builder class so that if it was called on a `DerivedBuilderForSpecificPurpose` instance it returns that instead of an instance of its own type? How about if someone else subclasses it as well - how does it know how to create an instance of every single possible subclassed type so it can be instantiated and returned?

**Comment by CyrusNajmabadi** [comment]
>Let's say you have IOrderedEnumerable...okay, so now how does LINQ method Where called on an IOrderedEnumerable return an IOrderedEnumerable instance?

Easy:

```c#
public static IOrderedEnumerable<T> Where<T>(this IOrderedEnumerable e, ...);
```

> How do you write the method in your BASE builder class so that if it was called on a DerivedBuilderForSpecificPurpose instance it returns that instead of an instance of its own type?

Easy

```c#
class class MyBuilder {
     MyBuilder SomeFluentCall();
}

class DerivedBuilder : MyBuilder {
    new DerivedBuilder SomeFluentCall();
}
```

**Comment by mikernet** [comment]
Okay, so what's wrong with what you have there? The compiler can't automatically generate that for you...it doesn't know how to translate a method that returns `MyBuilder` into one that returns `DerivedBuilder`.

**Comment by mikernet** [comment]
The reason it works for `this` is because it knows the type of `this`. It can't possible work for your scenario if any arbitrary instance can be returned because then it can't be upcast to the type that it was actually called on.

**Comment by CyrusNajmabadi** [comment]
>Okay, so what's wrong with what you have there? 

You're the one who made the claim: "that what you are asking is impossible"

I'm showing it's not impossible.  It's a very routine and common sort of fluent-building pattern.  You make chained fluent calls and depending on what you call, you get different shapes back (often subtypes), so that you have the core functionality but you also have new functionality on top of that.

**Comment by mikernet** [comment]
It's impossible for the compiler to automagically transform your method that returns an instance of the current type into every possible subclass of that type. You have to write that logic yourself. `this` can be cast to the type of `this`. The `IEnumerable<T>` that comes from the `Where` method can't be cast to `IOrderedEnumerable` so how the heck would that work?

**Comment by CyrusNajmabadi** [comment]
> It's impossible for the compiler to automagically transform your method that returns an instance of the current type into every possible subclass of that type

Why would the compiler have to do this?  The 'this' type is how the type-decl declares its contract, and how the caller perceives what they get back when they call the method on the actual value.  Compiler is free to insert any necessary casts at callpoint with the assumption that the type-decl didn't lie and that implementers are abiding by that.

**Comment by mikernet** [comment]
Okay, so now write me a `Where()` method that returns IEnumerable<T> if called on that, an `IOrderedEnumerable<T>` if called on that, a `List<T>` if called on that, etc. ad infinium.

So now the requirements for implementing an IEnumerable object is that you also have to add casting operators that convert to every single possible type that might come back from a LINQ operator? That's absurd.

**Comment by mikernet** [comment]
You've defined a method that returns `this`, but *it doesn't*. It returns a specific type of object, and now casts that can easily fail under various circumstances are magically inserted. That's just a terrible idea.

**Comment by CyrusNajmabadi** [comment]
> Okay, so now write me a Where() method that returns IEnumerable if called on that, an IOrderedEnumerable<T> if called on that, a List<T> if called on that, etc. ad infinium.

```c#
new IOrderedEnumerable Where(...)
...
new List<T> Where(...)
```

> You've defined a method that returns this, but it doesn't

It doesn't?  In what cases does it not?  Note: i'm not saying that "this isntance" is being returned.  I'm sayin "an instance *of this type*" is being returned.

>It returns a specific type of object, and now casts that can easily fail under various circumstances are magically inserted

When do they fail?

**Comment by mikernet** [comment]
Not a method signature, a *method*. Like, with a body and stuff. How do you write a `Where` method, that returns an instance of `this`, which could literally be anything that implements `IEnumerable`?

You would have to define casting operators from every single builder class to every single other builder class. Do you not see how that's a horrible situation? Yeah it might work for 2 builders, but any more than that and you have a flustercuck of a webby mess. How is this a reasonable feature?

**Comment by CyrusNajmabadi** [comment]
> Not a method signature, a method. Like, with a body and stuff. How do you write a Where method, that returns an instance of this

```c#
this Foo() {
    // any code i want.
    return this;
}
```

**Comment by mikernet** [comment]
You just returned `this`. This entire conversation got started because they want to return stuff other than `this`, just `this` typed.

**Comment by CyrusNajmabadi** [comment]
> You would have to define casting operators from every single builder class to every single other builder class. Do you not see how that's a horrible situation? Yeah it might work for 2 builders, but any more than that and you have a flustercuck of a webby mess. How is this a reasonable feature?

You don't need that at all.

**Comment by mikernet** [comment]
Okay, so show me a method that returns an instance that isn't `this` in a `this` return typed method. That's literally what they were talking about. We already solved the problem of returning `this`.

**Comment by CyrusNajmabadi** [comment]
>You just returned this. This entire conversation got started because they want to return stuff other than this, just this typed. We already have a solution for what you just wrote.

```c@
this Foo() {
    // any code i want.
    return CreateInstance();
}

abstract this CreateInstance();
```

**Comment by CyrusNajmabadi** [comment]
> Okay, so show me a method that returns an instance that isn't this in a this return typed method. That's literally what they were talking about. We already solved the problem of returning this

See above.  You put the onus (through a bog standard abstract/override) on the subtype to produce instances as appropriate.

**Comment by mikernet** [comment]
Right...so `IEnumerable.Where` is now implemented ... how? Every single collection now needs to have methods on it to create an instance of its own type and populate it with the results of the `Where` operation? 

I'm sorry I'm not continuing this any further. It's obvious to me why any real world application of this is extremely limited. You're giving me these limited examples with no concrete real world representation and with all due respect, I don't think you're considering how this blows up. Your toy example doesn't do anything, it just returns a created instance. If you have a `Where` method that returns its own type, now you need every single `IEnumerable` to create its own type that can be populated with the results of that `Where` operation. 

This is useless.

**Comment by mikernet** [comment]
What if I have an IEnumerable that just counts from 1 to infinity: `class Counter : IEnumerable<int>`. Now I need to add support to this class to return an instance of `Counter` when `Where` is called? Why would I want that?

**Comment by HaloFour** [comment]
What about it?  That's only one way through which contravariant return this types can be utilized, and it probably doesn't make sense for LINQ.  But that's just one example of fluent APIs where the return values are immutable.  You can apply this to any builder, where the builder returns a new instance of itself but using multiple interfaces to establish dialects.  Most of my builders follow a pattern like this.

**Comment by CyrusNajmabadi** [comment]
> Every single collection now needs to have methods on it to create an instance of its own type and populate it with the results of the Where operation?

Sure.  What's the problem with that.  If you want to state that your contract is "this method must return an instance of 'this' type", then yes, your type hierarchy must abide by that contract.

**Comment by CyrusNajmabadi** [comment]
>  Your toy example doesn't do anything, it just returns a created instance

You literally asked me how it could be done.  I demonstrated exactly how it could be done in a trivial fashion.  There's no need to bog things down wiht an overwrought example when it's easy to just so the basics on why what i'm saying is valid and that your complains that things are impossible is incorrect.

**Comment by CyrusNajmabadi** [comment]
>This is useless.

It's extremely useful.  It may require constraints on your inheritance hierarchy.  But... so what?  Why would that bother me.  If i say "this method must return instances of this type" then yes... that will require that be provided.  **That's the point**.

**Comment by HaloFour** [comment]
I don't get what the objection is.  It's not like it would prevent you from returning the same instance.  It would just also allow you to return other instances.  It would build on the same basic compiler features which would leverage contravariant returns in the CLR to emit bridge methods that would return the concrete type.

**Comment by mikernet** [comment]
And those builders work fine the way they are. I don't see what you are getting out of making things extremely complicated and unwieldy.

Instead of silly toy examples, let's go through one actual example. This is my last attempt to show you why it's useless. Give me one example that this would *actually* be beneficial and I will show you how convoluted it gets in *any* real world situation. So far every example you've given me hasn't been convincing in the least bit and we just come back to "well that's not the best example". Give me your best example.

**Comment by HaloFour** [comment]
@mikernet 

> And those builders work fine the way they are. I don't see what you are getting out of making things extremely complicated and unwieldy.

Except when you switch to a dialect interface that extends the base interface and invoke one of the methods on the base interface which returns a base result and you lose the dialect methods.

**Comment by CyrusNajmabadi** [comment]
You're the one claiming it can't be done, and stating things are impossible, when it's then trivial to show examples of how it's trivial to do.  I don't see why it's my job to go further when it's you that's making the claims.  If this can't be done, then we would have failed at these basic examples.  But it was literally just one liners that showed how simple it is to have this sort of pattern.  You're the one claiming it won't scale, and then saying the onus is on us to prove you wrong.  But your claims have already been proven wrong.  I don't see why there's any more need to continue just trying to convince you of this.

**Comment by mikernet** [comment]
I already explained what is impossible, and it's not what your examples were showing. If you just want to be able to create abstract `this` methods then that's fine, but that's not what it seemed you wanted originally. There's nothing in this proposal that stops that from being possible, I don't see how that's a concern.

**Comment by mikernet** [comment]
The original examples were talking about LINQ, about strings, and immutable builders that aren't subclassable and were just generally bad examples that gave off the wrong impression for what you wanted. If abstract `this` methods give you what you want then I guess this proposal works fine as is, I don't see what the issue is.

**Comment by mikernet** [comment]
What I'm trying to tell you is that any real-world attempt to use abstract `this` methods that actually do stuff, not just return empty types, in order to facilitate some kind of weird API that returns the original objects is going to get super convoluted super fast. I can't even picture what an actual useful API, let alone an implementation that uses this would actually look like and you haven't provided a real world example that I can demonstrate why it doesn't work.

**Comment by CyrusNajmabadi** [comment]
> about strings, and immutable builders that aren't subclassable

The examples there were simply to show that it's common to have fluent apis that don't return the actual "this" value.  They were explanatory.  You can certainly expand them out to understand how they would actually apply to non-sealed classes.

**Comment by CyrusNajmabadi** [comment]
> What I'm trying to tell you is that any real-world attempt to use abstract this methods that actually do stuff, not just return empty types, in order to facilitate some kind of weird API that returns the original object types is going to get super convoluted super fast.

Why?

> let alone an implementation that uses this would actually look like and you haven't provided a real world example that I can demonstrate why it doesn't work.

What part about the implementation are you not understanding how it would work? Examples have already been shown above how this could work at an impl level.  

>  If you just want to be able to create abstract this methods then that's fine

I'm really confused why you're having a problem with this then.  You seem to think this is both impossible, but you say that hte solutions i'm using to show how this is possible are fine as well.

**Comment by mikernet** [comment]
I take that back. Abstract `this` methods that can return anything other than `this` would not work without immense complications. I’ll explain when I get home.

**Comment by jnm2** [comment]
Immutable types typically return a new instance. My most frequent reason for wanting `this` types is immutable interfaces and base classes.

**Comment by mikernet** [comment]
These piecemeal examples do not demonstrate the whole picture as to why this won't work and why it is convoluted. You just keep pushing the problem one level back but the underlying issue is still there because we aren't going through a full example. Give me an example definition of a base class and a subclass, with methods that return this. Make the methods sensible so I know what they should be doing and I will demonstrate the problem.

**Comment by mikernet** [comment]
Here's a simple example to demonstrate one of the problems:
```c#
abstract class A {
    public this Foo() {    
        return CreateInstance();
    }

    public abstract this CreateInstance();
}

class B : A {
    public override this CreateInstance()
    {
       return ??  // What goes here ?!? "this" might be a C so we can't return a B.
       // How do we use a "this" variable so that we can transform it as needed here?
       // It can't be type B for the reason above. If it's type "this" then how do we get an instance
       // without endlessly calling virtual methods that don't actually have an implementation?
    }
}

class C : B {
   // Do we force this class to override CreateInstance even though the base class implemented it already?
}
```

This is just a sampling of problems. There are more. I don't think what you are trying to achieve is possible with automagic compiler behavior. There are already mechanisms to do what you are doing, namely with "new" methods, which give you the manual control necessary to actually accomplish the API you are looking for.

If you're still not convinced then I would be happy to go through and build a minimal but complete example with you that we can pick apart and demonstrate the problems on.

**Comment by CyrusNajmabadi** [comment]
> // What goes here ?!?

An instance of B.  Subtypes would be required to supply this to provide their own impl that supported the constraint.  As i said above: 

> It's extremely useful. It may require constraints on your inheritance hierarchy. But... so what? Why would that bother me. If i say "this method must return instances of this type" then yes... that will require that be provided. That's the point.

**Comment by CyrusNajmabadi** [comment]
> would not work without **immense** complications.

> The reason I'm pushing back on this is that it will add a **LOT** of complication

(emphasis mine)

I think the problem is that *you* have somehow internalized that this is some sort of a significant issue.  Nothing you've stated seems like any sort of challenge at all.  It would be fairly trivial to have some simple rules here on both how declarations work, and what requirements it places on subtypes.

**Comment by mikernet** [comment]
I'm saying it's convoluted with weird rules and it's very complex for little benefit. You can already do this with `new` without mangling the language with this weirdness. Based on the above, surely you realize that's never going to happen right?

**Comment by mikernet** [comment]
The proposal requires adding some syntactic sugar. This requires changing the meaning and rules around what abstract methods are in certain situations, checking base implementations based on whether they return `this` or an instance of some sort...which now means that changing the content of a method from `return this` to `return new B()` without changing the method signature at all breaks libraries that have subclasses of your objects...it's a bloody mess.

**Comment by CyrusNajmabadi** [comment]
> I'm saying it's convoluted with weird rules and it's very complex for little benefit.

You started by stating it was impossible.  It's not impossible.  it's very possible, whether or not you think the benefits are worth it is a *secondary* concern that comes down to opinion.  I personally think this would be valuable.  To address your points:

1. "weird rules". I don't see anything weird about the rules.
2. "it's very complex".  It doesn't seem complex at all to me.
3. "for little benefit".  that's fine if you believe that.  I think it's a reasonable benefit.  fluent apis that return different subtypes are totally reasonable IMO.

> You can already do this with new without mangling the language with this weirdness

I don't see anything weird here.

> Based on the above, surely you realize that's never going to happen right?

Given i don't agree with any of your points (and one is pure opinion)... no i don't.  But what do i know, i only was on the C# LDG for more than a decade :)

**Comment by jnm2** [comment]
@mikernet Changing an existing virtual method's return type to something more derived is already a bad breaking change. Changing it to `this` is the same old concept.

**Comment by CyrusNajmabadi** [comment]
> This  requires changing the meaning and rules around what abstract methods  are in certain situations

Sure.  it's a new feature. we can define waht it means to use that new feature.  Since it's a return-type feature (and would presumably be done with annotations), it seems pretty easy to do.

> checking base implementations based on  whether they return this or an instance of some sort...which now means that changing the content of a method from return this to return new B() without changing the method signature at all breaks libraries that have subclasses of your objects...

You are changing the method signature.  You're returning the "this-type".  That's already statement about what's going on.  

> it's a bloody mess.

It doesn't seem messy here.  I'm not sure it's worth doing.  But all my points here stemmed from your strong claims from the beginning that this wasn't possible, where it seems pretty simple to do.

**Comment by mikernet** [comment]
@jnm2 This isn't changing the return type, it's changing the contents of the body of the method to either `return this` or `return new Class()`. The return type has to be `this` in either case.

**Comment by jnm2** [comment]
I'm not seeing a breaking change 8n that example.

**Comment by mikernet** [comment]
The breaking change is that if you `return this` then subclasses don't need to implement the overridden abstract / virtual method. If you `return new ThisClass()` then they do.

**Comment by CyrusNajmabadi** [comment]
Yes... that woudl be a break... because you're breaking things.

**Comment by mikernet** [comment]
@CyrusNajmabadi  I don't know why you keep bringing up the impossible thing. I don't know how many times I need to say this...that was in response to you claiming that LINQ methods could use this, and based on that example, yes it would be "impossible", unless you want every single IEnumerable to implement methods to support every LINQ operation...which would never happen, thus impossible.

**Comment by jnm2** [comment]
@mikernet 

> The breaking change is that if you `return this` then subclasses don't need to implement the overridden abstract / virtual method. If you `return new ThisClass()` then they do.

This doesn't make sense to me. Returning `new ThisClass` would not be allowed in a non-sealed class. That would break the strongly-typed `this` contract.

**Comment by CyrusNajmabadi** [comment]
> unless you want every single IEnumerable to implement methods to support every LINQ operation...which would never happen, thus impossible.

That's totally reasonable to think about supporting.  If we had DIM from the start, we likely would have done that.    There's nothing impossible here.  It may not be done with *existing* baked APIs since they're already out there.  but that's a statement more about utilizing a new feature on a shipped API, versus how it would be written now if we had this feature, or how we would write future APIs.

**Comment by mikernet** [comment]
@jnm2 I tend to agree, but then the only types that could ever be returned from a `this` method would be sealed types or `this`. That seems rather limiting but maybe not. @CyrusNajmabadi made the suggestion about being able to return instances of the current class so I rolled with it.

**Comment by CyrusNajmabadi** [comment]
> This doesn't make sense to me. Returning new ThisClass would not be allowed in a non-sealed class. That would break the strongly-typed this contract.

i think it would be fine to provide.  BUt it would require subtypes must provide their own way to satisfy things.  In a real sense, this would be a language concept where there can be concrete types, but thye effectively have an abstract-method *only* for their subclasses.

**Comment by mikernet** [comment]
DIM wouldn't help you return an instance of every single IEnumerable implementing type. Not to mention, you wouldn't want to. The whole point of LINQ is to defer execution and to use enumerating state machines, not get back a chain of `List<T>`s. Nothing about that example made sense.

**Comment by jnm2** [comment]
@mikernet 

> @jnm2 I tend to agree, but then the only types that could ever be returned from a `this` method would be sealed types or `this`. That seems rather limiting but maybe not. @CyrusNajmabadi made the suggestion about being able to return instances of the current class so I rolled with it.

Believe me, what's limiting is not being able to do exactly those two things.

**Comment by CyrusNajmabadi** [comment]
> The whole point of LINQ is to defer execution and to use enumerating state machine

Nothing about returning a derived type *if you choose* means that you're non-deferred.

**Comment by SilentSin** [comment]
Here are some examples of the problems I see:

-----

We have the `ICloneable` interface.
```
interface ICloneable
{
    object Clone();
}
```
Wouldn't it be awesome if we could constrain that `object` to be a copy of the current object? Yes, it sure would:
```
interface ICloneable
{
    this Clone();
}
```
Now let's implement it:
```
class MyCloneableBaseClass : ICloneable
{
    public this Clone()
    {
        return new MyCloneableBaseClass();
    }
}
```
Looks great, but now I want to inherit from that:
```
class MyCloneableChildClass : MyCloneableBaseClass
{
}
```
And now we have a problem. `MyCloneableChildClass` can't override `Clone` because it's not `virtual`, but `Clone` isn't going to ever return a `MyCloneableChildClass`. So to actually implement a method with the `this` return type, we have a situation with 3 totally new unprecedented constraints:

- The method must always be `virtual` (or `abstract` in an `abstract` class).
- Despite being `virtual`, the method must always be overridden in every child class as if it were `abstract` and it can't be `sealed` at any point.
- Despite being `virtual` and having an actual implementation in `MyCloneableBaseClass`, the `MyCloneableChildClass` implementation of `Clone` shouldn't be allowed to call `base.Clone()` because while `return (MyCloneableChildClass)base.Clone();` looks like it should compile, it will literally never work.

That's an absolute ton of extra changes the language would need which @CyrusNajmabadi would tack onto this proposal even though none of them have actually been clearly elaborated until now. Least of all in the OP, which only talks about returning `this` itself. Those are the "weird rules" @mikernet is referring to. They could even potentially be useful on their own entirely separately from the concept of method chaining, yet you would shove them into what started out as a rather simple proposal.

-----

Now that we have those requirements, something simple that does actually just want to `return this` gets more complicated:
```
interface ILoggable
{
    this Log();
}

class MyLoggableBaseClass : ILoggable
{
    private float x, y, z;

    public virtual this Log()
    {
        Debug.Log(x);
        Debug.Log(y);
        Debug.Log(z);
        return this;
    }
}

class MyLoggableChildClass : MyLoggableBaseClass
{
    public override this Log()
    {
        base.Log();
        return this;
    }
}
```
Those requirements force us to create an override in the child class which does practically nothing new. This feature is now even more tedious than having multiple constructors because you always need an implementation in every child class.

-----

So in one example, calling the base method will never be correct because it creates an invalid clone while in the other example all the child implementation actually does is call the base method. Basically, any time the base method actually has `return this;` it will be safe (and probably expected) for the child to call it, but any time it creates a new instance it will most likely be totally wrong for the child to call it. Yet the language gives us no real way of distinguishing between the two when implementing a child class.

Alternatively, if we take a step back to the OP where the `this` return type literally only needs to mean `returns this`, then none of these complications are necessary.

I've been trying to think of another example that would somehow return an existing instance of the current type other than `this` but haven't been able to come up with one. So it seems like all these extra rules can only ever be useful for methods that create a new object. So if there are only two situations where it can work which both have notably different requirements, I'd say this should be two separate features rather than sticking them both in one that has to cripple one side:

1. Methods that want to `return this` don't need any additional rules. Having `this` as the return type can implicitly add `return this` to the method and the compiler can automatically generate methods for any child class which look like `MyLoggableChildClass.Log`. This is what I would want the feature for and I would hate to have to manually put those identical methods in every child class.
2. Methods that do not want to `return this` do need the additional rules and thus should have a separate keyword. `instancetype` sounds like it would fit the bill nicely if that's what Objective-C uses it for. But this capability should not overcomplicate situations where all we want is to `return this`.

**Comment by mikernet** [comment]
Yes I'm aware. That still doesn't explain how this is actually useful in LINQ and what you mean by `if you choose` is clear as mud. Regardless, I've lost interest in trying to figure it out. Without a fleshed out example of your use case this is going absolutely nowhere, we're just talking in piecemeal abstracts and I don't have an actual example that makes sense to critique, so I'm bowing out until that happens (if ever). 

The changes required for this in comparison requires some pretty deep changes to the runtime that aren't required for the original proposal. It is worlds apart apart in terms of effort required to implement and design from the scope of the original proposal. As a 10 year veteran of the C# LDG I would presume you can see that ;)

**Comment by CyrusNajmabadi** [comment]
> The changes required for this in comparison requires some pretty deep changes to the runtime that aren't required for the original proposal.

I see nothing about this currently that would require *any* runtime changes at all...

**Comment by jnm2** [comment]
@SilentSin This looks invalid to me. `new MyCloneableBaseClass()` can't satisfy the return type `this` unless `MyCloneableBaseClass` is sealed.

> Now let's implement it:
> 
> ```
> class MyCloneableBaseClass : ICloneable
> {
>     public this Clone()
>     {
>         return new MyCloneableBaseClass();
>     }
> }
> ```

**Comment by mikernet** [comment]
@CyrusNajmabadi The rules around validating bytecode, for one...because now some virtual methods *require* overrides in all sealed classes, among the other rules that @SilentSin explained. This must be done at runtime because assemblies can be updated independently.

**Comment by jnm2** [comment]
@mikernet That's a false example. Virtual methods would only be allowed to return `this` or the return value of a `this`-returning method.

**Comment by CyrusNajmabadi** [comment]
> And now we have a problem.  ...
>
> That's an absolute ton of extra changes the language would need which @CyrusNajmabadi would tack onto this proposal even though none of them have actually been clearly elaborated until now

I'm really failing to see how words like "immense", "ton" and "lot" apply here.  They seem fairly mundane and easy.  You want to use the 'this'-type, there will be some restrictions placed on your class that subtypes will have to abide by.  For structs/sealed-classes this is of course not a problem.  For normal classes, it doesn't even seem like a difficult thing to provide/explain.  If you subclassed and there was a method that then required you to now provide an impl for, you'd have to provide it.  In that sense it's similar to 'abstract' (though i'm not saying you'd actually use abstract methods for this).

IN terms of your direct points.

> And now we have a problem. MyCloneableChildClass can't override Clone because it's not virtual

It's quite possible that using the 'this' type in a non-sealed type, and not returning 'this' would require the method to be `virtual` *precisely* so that subclasses could ensure they provided the right result.  This could even be implicit (though i would likely personally be against that due to how the language works today).

> Despite being virtual, the method must always be overridden in every child class as if it were abstract and it can't be sealed at any point.

Yes... that seems trivial.  

> Despite being virtual and having an actual implementation in MyCloneableBaseClass, the MyCloneableChildClass implementation of Clone shouldn't be allowed to call base.Clone()

I'm not certain why that would be the case.  If you're calling through 'base.Clone()' then the 'this' type in that sense in 'base'.  So you would see base.Clone() returning `MyCloneableBaseClass`. 

None of this seems particularly difficult.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi The rules around validating bytecode, for one...because now some virtual methods require overrides in all sealed classes, 

They're not required from a runtime sense.  No more than a virtual `string` return type in one assembly will cause something to fail to validate if it returns `string?` as an override in a different assembly.  These can be entirely on top of the runtime.

**Comment by jnm2** [comment]
Ah, looks like @CyrusNajmabadi is going a different direction with this than I'm imagining. I like sealed classes, and I'm really not interested in anything except sealed and abstract classes.

**Comment by CyrusNajmabadi** [comment]
>Now that we have those requirements, something simple that does actually just want to return this gets more complicated:

 >Those requirements force us to create an override in the child class which does practically nothing new. This feature is now even more tedious than having multiple constructors because you always need an implementation in every child class.

That is not hte case.  Your original method was: 

```c#
    public virtual this Log()
    {
        Debug.Log(x);
        Debug.Log(y);
        Debug.Log(z);
        return this;
    }
```

You would never have to override this because this would be a `this-method` that doesn't need subclasses to be involved since you know it's always returning something of hte right type.

**Comment by CyrusNajmabadi** [comment]
> Ah, looks like @CyrusNajmabadi is going a different direction with this than I'm imagining. I like sealed classes, and I'm really not interested in anything except sealed and abstract classes in cases where I want this typing.

Yes.  That's another approach (an probably a cleaner and better one to suggest people use).  I just don't see it as mandatory, though i would personally advise that sort of coding pattern.

**Comment by HaloFour** [comment]
@mikernet 

IL has no problem with contravariant returns.  No runtime changes are required for C# to add that as a feature.  To the CLR a method signature includes the return type, so another method with identical parameters and a different return type is just an overload.  Contravariant returns are another proposal here.  "this-type" could just piggy back off of that by automatically generating the contravariant returns and necessary bridge methods.

**Comment by CyrusNajmabadi** [comment]
> Methods that want to return this don't need any additional rules. Having this as the return type can implicitly add return this to the method and the compiler can automatically generate methods for any child class which look like MyLoggableChildClass.Log. This is what I would want the feature for and I would hate to have to manually put those identical methods in every child class.

In terms of ergonomics, i would be totally ok with this.  There's an open question to me about how you would code the *body* of the 'this-method'.  And how this would appear at the **metadata** layer (i.e. for pre C#vnext compilers).   At the metadata level, this could just be the top level method, and the compiler could know it would need to cast at callsites.  Or, it could mean the compiler inserted the downlevel methods with the right return-type per subclass.    

However, while interesting topics, these are of slightly less interest to me.

**Comment by SilentSin** [comment]
@jnm2 That's my point. Either `MyCloneableBaseClass` must be `abstract` to avoid providing an implementation (in which case every type that inherits must also be `abstract` until the final child which must be `sealed`) or we need these extra rules that require every child class to provide its own new implementation anyway.

@CyrusNajmabadi 

> You want to use the 'this'-type, there will be some restrictions placed on your class that subtypes will have to abide by.

Wrong. I want to use the `this`-reference specifically, which does not require those same restrictions.

> Yes... that seems trivial.

Trivial, but more importantly unnecessary and extremely annoying. All I want the feature for is methods that `return this`, yet you would have me implement the same `base.Method(); return this;` for every one of those in every child class.

> I'm not certain why that would be the case. If you're calling through 'base.Clone()' then the 'this' type in that sense in 'base'. So you would see base.Clone() returning MyCloneableBaseClass.

But the `base.this` pointer is actually a `MyCloneableChildClass` so a proper valid clone of it must also be a `MyCloneableChildClass`, which is not what `base.Clone` creates.

> You would never have to override this because this would be a this-method that doesn't need subclasses to be involved since you know it's always returning something of hte right type.

But you can't tell that from the method signature, hence why I suggested splitting it into return type `this` for the original proposal and return type `instancetype` for what you are adding to it.

**Comment by CyrusNajmabadi** [comment]
> Contravariant returns are another proposal here. "this-type" could just piggy back off of that by automatically generating the contravariant returns and necessary bridge methods.

I believe the compiler already does this in several cases.  I think even from an impl perspective this would be easy to accomplish.

**Comment by mikernet** [comment]
This still doesn't clear up how you would actually write a method that does something useful and I can't really critique that without an actual example.

**Comment by CyrusNajmabadi** [comment]
> Trivial, but more importantly unnecessary and extremely annoying. All I want the feature for is methods that return this, yet you would have me implement the same base.Method(); return this; for every one of those in every child class.

No i wouldn't.  Where are you getting that from.  I've never ever said that.

**Comment by CyrusNajmabadi** [comment]
> But the base.this pointer is actually a MyCloneableChildClass so a proper valid clone of it must also be a MyCloneableChildClass, which is not what base.Clone creates.

The language can say whatever it wants here.  When invoking a this-returning-method on 'base' it can state whatever it wants about the 'this-type' in that signature.

**Comment by HaloFour** [comment]
I'm not even opposed to the notion of it returning `this` by default (although I'm not sure I like the idea of an implicit return in any case).  I just don't see why the implementation would close the door on supporting returning another instance that satisfies the requirements of also being of the same type as `this`.  It adds more flexibility to the feature.  If contravariant returns weren't a potential thing and the CLR raised barriers to implementing them maybe I would side on making this behave more like method cascades, but as someone who prefers immutability to my fluent APIs it immediately means that I wouldn't benefit.

**Comment by CyrusNajmabadi** [comment]
> But you can't tell that from the method signature,

Sure you can.  This would presumably be encoded with attributes (since the runtime has no such concept here).  So the attribute can trivially state what's going on here...

**Comment by HaloFour** [comment]
Runtime can't tell `dynamic` or NRT from the signature either.

**Comment by jnm2** [comment]
@CyrusNajmabadi I think there's some confusion stemming from talking about two scenarios with different characteristics: returning `this`, which doesn't require newing and therefore doesn't require all classes to be abstract or sealed, and returning a new instance (which does require all newing to be pushed to sealed classes, with the rest abstract).

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I think there's some confusion stemming from talking about two scenarios with different characteristics: returning this, which doesn't require newing and therefore doesn't require all classes to be abstract or sealed, and returning a new instance (which does require all newing to be pushed to sealed classes, with the rest abstract).

Right, but i called that out explicitly:

> Yes, most likely. If your impl returned 'this' (or a this-type), then the method woudl be marked as not needing to be overriden. But if it returned a different instance, then it would be marked as something the subclass would have to supply the impl for.

> In more concrete terms, the compiler would spit out bridges automatically depending on what was going on. If you didn't return 'this' (or a this-type), then abstract bridges would have to be there to force the subtype's involvement.

**Comment by SilentSin** [comment]
> Sure you can. This would presumably be encoded with attributes (since the runtime has no such concept here). So the attribute can trivially state what's going on here...

So now when I'm implementing a child class containing such a method I need to go to the definition to look at the attributes to figure out how it's supposed to work (unless we also change Visual Studio to show attributes in the tooltip).

Is there any other situation where you currently need to look at the attributes of a base method before attempting to override it?

You still haven't acknowledged the idea of using a separate `instancetype` keyword, but if your only objection is the usual "adding new keyword = bad" then it could be done with existing keywords. For example, `new this` would nicely indicate that it isn't simply returning `this` (and therefore must always be overridden in every child class), or alternatively `implicit this` would nicely indicate that it is actually just implicitly returning `this` (and therefore doesn't need to be overridden).

**Comment by mikernet** [comment]
I'm trying reaaaaally hard to write a sensible 3 class "immutable builder" API that makes sense (one base class, two subclasses) and I'm struggling to come up with any reasonable API that can actually make use of this feature. At the very least, it seems to me that casting to `this` and a `this` typed variable would be required to make anything workable, and then I still can't get a sensible implementation of an immutable builder hierarchy going. 

Can someone smarter than me please put up a 3 class sample of how this all works?

**Comment by CyrusNajmabadi** [comment]
> So now when I'm implementing a child class containing such a method I need to go to the definition to look at the attributes to figure out how it's supposed to work (unless we also change Visual Studio to show attributes in the tooltip).

Seems like a reasonable strategy.  

> You still haven't acknowledged the idea of using a separate instancetype keyword,  but if your only objection is the usual "adding new keyword = bad"

I have placed no objection on it.  It doesn't seem necessary to me.  But i'm not opposed to it.  Its presence or absence doesn't really change anything for me.  If LDM went either direction i'd be fine with it.

**Comment by CyrusNajmabadi** [comment]
Note: IMO, i would also be fine with a restricted for of `this`-types that required the result to be `return this;` (or returning another `this`-type expression).  It could always be relaxed in the future.  I just don't see anything fundamentally blocking this supporting the broader case.

**Comment by SilentSin** [comment]
> I just don't see anything fundamentally blocking this supporting the broader case.

I don't either, but the attitude I've seen from you and others in the past is that changes requiring new concepts or blurring the definitions of existing concepts have a much higher bar to meet for acceptance.

Allowing an implicit `return this` with the actual return type being whatever the current child type is would not be much of a departure from existing concepts like `var`.

However, a system where methods are a hybrid of abstract and virtual where they have an implementation but also need to be overridden in every child class unless they `return this` isn't really similar to anything the language currently does. It would require more effort all around, from the compiler implementation to learning to use it. It might potentially be useful, but that's not what I want the feature for and I would hate to see the whole thing get denied due to this additional complexity.

**Comment by mikernet** [comment]
If every feature starts at -100, then the more complicated version of this feature which changes fundamentally engrained concepts and WILL initially cause confusion and many "why doesn't this work? It seems like it should" and "Why am I getting this error when all I'm trying to do is inherit this class? I've never had this problem before" questions probably starts at -1000. Muddying the waters about the meaning of a virtual/abstract method has significant additional cost and places an additional cognitive load on the language, particularly for beginners. The reason I keep pushing for an actual "real world" example is that I can't see how this would overcome that initial -1000 and I've tried really hard.

I think it is noteworthy that you can always add this additional capability later if the need for it is shown to be real.

As an alternative to writing out an example from scratch, how about this: point me to an existing library with an immutable builder that could have an improved API surface if this feature was available and give me one concrete example of how that improvement would look. If it makes sense, I'll go ahead and write what a sample implementation for that change would look like if this feature was available and let you know the problems I run into, because I'm sure there will be several.

**Comment by CyrusNajmabadi** [comment]
> I don't either, but the attitude I've seen from you and others in the past is that changes requiring new concepts or blurring the definitions of existing concepts have a much higher bar to meet for acceptance.

I don't think you understand my position :)  I'm not arguing that it should be done.  I'm arguing that certain points against it were incorrect.  That's not an endorsement of it actually being done as a feature.  Just an attempt to fairly assess what is possible and what isn't.

**Comment by CyrusNajmabadi** [comment]
> If every feature starts at -100, then the more complicated version

I think of the argument in the reverse.  Because you're at a negative, you have to prove more value ot be taken.  So a narrower focused feature (like just `return this;`) may not reach that level.  But something more broadly applicable might.

That said, as i mentioned [here](https://github.com/dotnet/csharplang/issues/902#issuecomment-487257163), my argument has never been "the feature must support returning something other than `this`"  Rather, it has been: i don't see any core problem that woudl stop the feature from working with things other than `return this;`

**Comment by CyrusNajmabadi** [comment]
> "Why am I getting this error when all I'm trying to do is inherit this class?"

The same reason you get other sorts of errors when inheriting: Because your base class has placed constraints on you that you need to provide.  

Having a new feature add new constraints doesn't bother me.  That happens all the time.  That people can and will run into those constraints isn't something out of the norm for a language feature.

**Comment by SilentSin** [comment]
> i don't see any core problem that woudl stop the feature from working with things other than return this;

The core problem is that it can't be done with the same feature.

An implicit `return this` could be implemented as some simple syntax sugar on the caller side and/or having the compiler generate `new` methods of the appropriate type in each child class which simple call `base.Method(); return this;`.

But implementing that doesn't seem like it would make any noteworthy contribution towards the implementation of a system requiring methods that both have an implementation and must always be overridden and the overrides change the return type to match each child class.

I've worked on two different APIs recently that would benefit greatly from the former, but I can't think of any real example where the latter would provide much benefit to me (and evidently no one else can either because we still haven't seen a proper example). Thus, I think it is very important to keep the simpler and more useful feature entirely separate from the more complex and less useful feature.

**Comment by mikernet** [comment]
> Note: IMO, i would also be fine with a restricted for of `this`-types that required the result to be `return this;` (or returning another `this`-type expression). It could always be relaxed in the future. I just don't see anything fundamentally blocking this supporting the broader case.

The blocking factor is the fact that no real use cases have been demonstrated and there is a significant additional cost to implementing the broader feature.

**Comment by CyrusNajmabadi** [comment]
>  and there is a significant additional cost to implementing the broader feature

You keep saying things like "lots" or "immense" or "significant", and i can't fathom why.  None of this is particularly complicated or difficult to spec out/implement.

**Comment by CyrusNajmabadi** [comment]
> Thus, I think it is very important to keep the simpler and more useful feature entirely separate from the more complex and less useful feature.

That's fine.  That's a fundamentally different argument than the one i was responded.  As i've already stated *numerous* times already, i would be totally ok with this being scoped down.  I'm not sure how this is being lost in our conversation.

My only point has been that there's nothing particular difficult about supporting the broader case.  That wasn't an endorsement that it *should* be done, just a counterpoint to those saying it could *not* be done.  

Whether something shoudl/shouldn't be done comes down to personal opinion on the importance/weight of things.  Whether things can/can't be done come down to actual practical concerns relatin to what's possible in the language spec and what can be done at the implementation level.  I have only been discussing the latter (the can/can't part).  I'm personally fine with any opinion on should/shouldn't.  And *as i've been explicit on*, i would most likely go with a scoped-down version initially.

**Comment by mikernet** [comment]
What’s difficult is implementing this in a way that is useful and achieves some kind of goal. So far, based on our discussions, I find it entirely unconvincing that this can be used in a useful manner the way others seem to think it can be used. 

Can it be implemented in a way that makes it completely useless? Sure, I guess, if that’s what you’re going for, but I don’t see the point of having an academic discussion about something with no practical applications. When I originally said what your were trying to achieve with LINQ was “impossible”, I meant impossible to implement in a manner that is useful and could actually be used for the stated purpose in a reasonable fashion. I think that much should have been apparent.

Regardless of how easy you think this is to spec and implement, we are all wasting our time discussing a feature that requires some amount of effort to spec / implement and which nobody has demonstrated a useful purpose for, so I suggest we shelve further discussions to prevent derailing the original proposal any more unless and until such a purpose has been presented. I wasted a couple hours yesterday trying every which way to design a class structure that uses this feature to try to understand why you guys interested in investing effort into this to no avail and nobody seems interested in demonstrating a clear example so this has been a complete waste of time so far.

**Comment by mikernet** [comment]
I think step 1 of proposing any feature, regardless of the perceived “easyness” of it, is to show the need for it. We seem to have completely skipped over that part, so how about we rewind and take care of that first.

**Comment by HaloFour** [comment]
@mikernet 

The question is whether or not the LDM finds it compelling.  They already find contravariant returns compelling, enough to have championed it and drawn up a proposal.  That already puts it a step ahead of this proposal which doesn't have the attention of the LDM at all.  It'd be relatively simple to build this proposal on top of contravariant returns as a form of "type-this" returns on top of which fluent `this` APIs could be designed relatively easily.  But if this proposal doesn't build on contravariant returns and instead builds on top of cascades (which are not championed) it starts off from scratch as a completely orthogonal concern.  If contravariant returns are implemented without any form of "this-typing" support that may be considered enough to "scratch the itch" since it makes it a little easier to at least implement such APIs, even if you're still doing it manually.  I'd rather piggyback off of the work in progress and enable both scenarios.

**Comment by mikernet** [comment]
@HaloFour That's because contravariant returns have demonstrated use cases. So once again, if you can demonstrate a use case for this, I'm all on board.

**Comment by mikernet** [comment]
You keep talking about enabling scenarios without actually demonstrating a scenario that can be enabled by this. Please for the love of all that is good and holy just provide an example as requested and we can move past this banter.

**Comment by HaloFour** [comment]
@mikernet 

> That's because contravariant returns have demonstrated use cases.

And this proposal is a specialized case of contravariant returns.  Even if it is hacked via method cascades that is still what it is attempting to accomplish.

**Comment by HaloFour** [comment]
@mikernet 

> So once again, if you can demonstrate a use case for this, I'm all on board.

I'm not particularly interested in whether or not you're on board.

"Withers" are a good example of a use case for "this-type" returns.  They are required to both return a new instance as well as return the most specific type.  They can't be emulated via shadowing otherwise you get decapitation.  "Withers" are also championed as a part of the records proposal.  Immutable builders for specializable classes are basically the same thing, and I use them to construct message types which can be specialized.  Literally defined one in Java yesterday, although with contravaraint returns in Java I could at least emulate them with an override.

**Comment by HaloFour** [comment]
D'oh, got my "contravariant" and "covariant" backwards.  Thanks @CyrusNajmabadi.

**Comment by mikernet** [comment]
> I'm not particularly interested in whether or not you're on board.

Okay well then there is nothing more for me to discuss here. You can say things like "withers" and "constructing message types" and "immutable builders" all you want but the very clear problem with how to actually *implement* them using what is being proposed here is impossible for me to demonstrate without you actually **showing** an example of how it would be implemented in C# with this feature that can be dissected.

I wish you the best of luck in getting your feature adopted. Enjoy the rest of your weekend.

**Comment by CyrusNajmabadi** [comment]
> You can say things like "withers" and "constructing message types" and "immutable builders" all you want but the problem I can see clearly in my mind with how to actually implement them using what is being proposed here is impossible for me to demonstrate

Perhaps you should take your concerns then to the respective proposals?

**Comment by mikernet** [comment]
@CyrusNajmabadi Which respective proposal? If the matter of having `this-returning` types that can return objects other than `this` is being discussed somewhere else I would love to. Can you point me in the right direction?

**Comment by CyrusNajmabadi** [comment]
@mikernet There are a bunch out there.  Maybe try looking for builders/immutable-objects/withers/records/etc.  You could even look at the ones about covariant return types.  Though i don't particularly expect any individual proposal to be done, it will be more likely that several will be bucketed into a full feature/pillar for a future release that are all about the general theme.

**Comment by CyrusNajmabadi** [comment]
While i can't say for certain, this does feel like an area that the LDM may be very interested in lookin at once NRT is done.  Open question though if they'll eventually go here, or if they'll go in other big-pillar areas (like 'shapes').  But there's been enough interested in the pieces (and the whole) here over the year to give me a strong belief that something like this could have a reasonable chance of happening in the next 1-2 releases.

**Comment by mikernet** [comment]
I've been mostly keeping up with those proposals and after some initial searching I can't seem to find anything that would cause me concern on those proposals so if you or anyone else comes across the place where it is being discussed and posts a link to it here so the discussions can be cross-referenced that would be lovely.

**Comment by CyrusNajmabadi** [comment]
i believe in all the immutability proposals you'll be getting new instances.  But, for the case of records/withers, you'd want the resultant type to be your containing type.

**Comment by mikernet** [comment]
I just thought of another use case for a `this` type: strongly typed senders for events in a class hierarchy.

Specifically:

```c#
public delegate EventHandler<TSender, TEventArgs>(TSender sender, TEventArgs e);

public class Control {
    public event EventHandler<this, MouseEventArgs> MouseDown;
    //instead of:
    public event EventHandler<Control, MouseEventArgs> MouseDown;
}

public class Button : Control {
}

public class SomeWindow {
    private Button someButton;
    public void Run() {
        someButton.MouseDown += (s, e) => { }; // s is of type Button, not Control
    }
}
```

**Comment by YairHalberstadt** [comment]
@mikernet

What would the code gen be

Note this is already sort of possible with the following pattern:

public class Control<T> where T : Control<T>
{
    public event EventHandler<T, MouseEventArgs> MouseDown;
}

public class Button : Control<Button> {
}

public class SomeWindow {
    private Button someButton;
    public void Run() {
        someButton.MouseDown += (s, e) => { }; // s is of type Button, not Control
    }
}

**Comment by mikernet** [comment]
That pattern is not generally applicable to deeper hierarchies, which are particularly common with controls. You can do the same thing for getting a "this" return type for one level deep as well :)

Not sure about codegen. I presume similar to whatever solution is created for a `this` return type, i.e. creating a `new` event that hides the base event and does the cast in its `add` and `remove` accessor, or the caller does the cast with a generated delegate.

**Comment by mikernet** [comment]
Also that pattern doesn't really work here (or in many other scenarios as well) because now there's no way to treat `Button` like a base `Control` object and still have access to the event handler that way, i.e.:

```c#

Control control = GetControl(criteria);
control.MouseDown += OnControlMouseDown; //fails because no way to define the event on non-generic Control class

```

**Comment by VanKrock** [comment]
I found solution of my problem. We can use implicit interface implementation and extension methods.
``` csharp
    public interface IMyClass
    {
        void SomeChange();
    }

    public static class Extensions
    {
        public static T SomeChange<T>(this T obj) where T : IMyClass
        {
            obj.SomeChange();
            return obj;
        }
    }

    public class MyClass : IMyClass
    {
        void IMyClass.SomeChange()
        {
            Console.WriteLine("Some change");
        }
    }

    //We can override this method in derived class
    public class Derived : MyClass, IMyClass
    {
        void IMyClass.SomeChange()
        {
            Console.WriteLine("Derived change");
        }
    }
```

**Comment by mikernet** [comment]
Didn't you already present that solution in the first post of this issue?

**Comment by VanKrock** [comment]
> Didn't you already present that solution in the first post of this issue?

No. Examples in first post have a problems. Current solution resolve that problems. We can use private fields of `MyClass` and we can inherit `MyClass`. This example completely solves my problems with fully fluent interfaces.

**Comment by mikernet** [comment]
Oh I see what you mean. I do something similar - an extension method calls an internal virtual method on the class which can be overridden by derived classes, so interfaces arent really needed to make this work. The extension method can make discoverability a problem at times though and it's somewhat annoying to not have everything encapsulated.
